docs/julia.rst

```diff
@@ -0,0 +1,450 @@
+
+JULIA is an intermediate language that can compile to various different backends
+(EVM 1.0, EVM 1.5 and eWASM are planned).
+Because of that, it is designed to be as featureless as possible.
+It can already be used for "inline assembly" inside Solidity and
+future versions of the Solidity compiler will even use JULIA as intermediate
+language. It should also be easy to build high-level optimizer stages for JULIA.
+
+The core components of JULIA are functions, blocks, variables, literals,
+for-loops, switch-statements, expressions and assignments to variables.
+
+JULIA is typed, both variables and literals must specify the type with postfix
+notation. The supported types are ``bool``, ``u8``, ``s8``, ``u32``, ``s32``,
+``u64``, ``s64``, ``u128``, ``s128``, ``u256`` and ``s256``.
+
+JULIA in itself does not even provide operators. If the EVM is targeted,
+opcodes will be available as built-in functions, but they can be reimplemented
+if the backend changes. For a list of mandatory built-in functions, see the section below.
+
+The following example program assumes that the EVM opcodes ``mul``, ``div``
+and ``mod`` are available either natively or as functions and computes exponentiation.
+
+.. code::
+    {
+        function power(base:u256, exponent:u256) -> result:u256
+        {
+            switch exponent
+            case 0:u256: { result := 1:u256 }
+            case 1:u256: { result := base }
+            default:
+            {
+                result := power(mul(base, base), div(exponent, 2:u256))
+                switch mod(exponent, 2:u256)
+                    case 1:u256: { result := mul(base, result) }
+            }
+        }
+    }
+
+It is also possible to implement the same function using a for-loop
+instead of recursion. Here, we need the EVM opcodes ``lt`` (less-than)
+and ``add`` to be available.
+
+.. code::
+    {
+        function power(base:u256, exponent:u256) -> result:u256
+        {
+            result := 1:u256
+            for { let i := 0:u256 } lt(i, exponent) { i := add(i, 1:u256) }
+            {
+                result := mul(result, base)
+            }
+        }
+    }
+
+Specification of JULIA
+======================
+
+JULIA code is described in this chapter. JULIA code is usually placed into a JULIA object, which is described in the following chapter.
+
+Grammar::
+
+    Block = '{' Statement* '}'
+    Statement =
+        Block |
+        FunctionDefinition |
+        VariableDeclaration |
+        Assignment |
+        Expression |
+        Switch |
+        ForLoop |
+        BreakContinue
+    FunctionDefinition =
+        'function' Identifier '(' TypedIdentifierList? ')'
+        ( '->' TypedIdentifierList )? Block
+    VariableDeclaration =
+        'let' TypedIdentifierList ( ':=' Expression )?
+    Assignment =
+        IdentifierList ':=' Expression
+    Expression =
+        FunctionCall | Identifier | Literal
+    Switch =
+        'switch' Expression Case* ( 'default' ':' Block )?
```

Comment:
This is one way to create unreachable code, I would say. I don't think we should create a special case for bool.